[RF-Docs] [ci-skip] Active Job Basics guide (#57101)
rails/rails: guides/source/active_job_basics.md

Reviewer: This is a big example. The reader could use some help with what to look for.
Maybe clarify what to look for in the describing paragraph or add a comment in the code:
```suggestion
    # Use the serialized payment_token and billing_profile_id
    # when resuming at the following step.
    step :submit_enrollment do
```

The goal of background jobs is to move long-running or non-critical work out of the HTTP request-response cycle and into a background queue (such as the default Solid Queue), and keep the web requests fast and responsive. This separation allows applications to perform work asynchronously, scale background processing independently, and execute multiple tasks in parallel without blocking user interactions.

NOTE: Active Job is an interface, not a background job execution engine. It does not run jobs on its own, it needs a persistent queuing backend (such as Solid Queue) to process jobs reliably in the background.

Creating Jobs
-------------

One of the most common jobs in a modern web application is sending emails to users. Active Job can do this outside of the request-response cycle, so the user doesn't have to wait on it. Active Job is integrated with Action Mailer so you can easily send emails asynchronously:

```ruby
# If you want to send the email now use #deliver_now
UserMailer.welcome(@user).deliver_now

# If you want to send the email asynchronously use #deliver_later
UserMailer.welcome(@user).deliver_later

> ```

Author: Added commas

> This section provides a step-by-step guide for defining a job Ruby class and then using the `perform_*` method to enqueue work to be executed in the background.

Author: Done. Moved the mailer example and added an explanation of the difference.

What is Active Job?
-------------------

The Active Job Rails framework allows you to declare background jobs and execute
them on a queuing backend. It provides a consistent, high-level interface for
common asynchronous tasks such as sending emails, processing data, generating

> reports, and running periodic maintenance work like data clean-up or billing.

Reviewer 2: Opinion: How about if we merge the NOTE section into the introduction as the NOTE sections seems repeated ?

```suggestion
The Active Job Rails framework provides an interface to declare background jobs and execute them on a queuing backend (such as Solid Queue). It provides a consistent, high-level interface for common asynchronous tasks such as sending emails, processing data, generating reports, and running periodic maintenance work like data clean-up or billing.
```

Author: Sure, updated.

end
```

Note that you can define the `perform` method inside a job class with as many
arguments as you want.

If your application uses a custom abstract job base class instead of
`ApplicationJob`, you can use the `--parent` option with the generator. The
parent class must itself inherit from `ApplicationJob`. This can be useful for
grouping related functionality in one place.

For example, given a custom abstract job class using
[queue_as](https://api.rubyonrails.org/classes/ActiveJob/QueueName/ClassMethods.html#method-i-queue_as):

```ruby
class PaymentJob < ApplicationJob
  queue_as :payments
end
```

You can generate a new job that inherits from it:

```bash

> bin/rails generate job process_payment --parent=payment_job

Reviewer: ```suggestion
$ bin/rails generate job process_payment --parent=payment_job
```

```

The above creates a class that will use the `payments` queue:

```ruby
class ProcessPaymentJob < PaymentJob
  def perform(*args)
    # Do something later, uses the "payments" queue
  end
end
```

### Calling the `perform_*` Methods

Once you have defined a job class with a `perform` method, you'd typically call
it using [`perform_later`][] to enqueue the work to be executed on a queuing
backend. Or use [`perform_now`][] if you want the job to execute immediately
without queueing. Both `perform_later` and `perform_now` call `perform` under
the hood.

In the examples below, the methods can be called from anywhere in your Rails
application, most commonly from controllers, models, or other jobs.

To run a job immediately without enqueuing it:

```ruby
GuestsCleanupJob.perform_now(guest)
```

To enqueue a job to be performed later:

```ruby
GuestsCleanupJob.perform_later(guest)
```

> To enqueue a job to be performed tomorrow at noon:

Reviewer: What do you think about moving the `:wait` and `:wait_until` examples to the "Enqueuing Jobs" section, maybe in a "Enqueuing Later" (or something) section? The "Enqueuing Jobs" section already has other examples of using `set` (to set the queue name or priority).

Author: So looking at the other `set` example in the Enqueuing Jobs secation vs. the. These are about _when_ to call `perform` vs. the other one is about which queue/priority. So I like the idea of keeping these two examples here. However, I added a link to the `set` method here for reference as it's the first mention (and it has 4 different use cases actually). Let me know if this is better or if you feel there is a reason to move these still :)

Reviewer: I found the 5 single line explanations with single line examples somewhat off putting.
Explaining the `set` method breaks this and makes it easier on the eyes for me. ❤️ 
It also makes it clearer what how these examples differ from the default.

Author: Good to know 😊 I reformatted the code examples to put those lines inside the code block as comments. Let me know if you like it that way or want me to put it back. 

I think the `set` examples are relevant here so useful to have for the reader but totally open to how we format, no strong preference! So let me know if you want me change anything else in this section so it flows and is not off putting :)

users. Active Job can do this outside of the request-response cycle, so the user
doesn't have to wait on it. Active Job is integrated with Action Mailer so you
can easily send emails asynchronously:

```ruby
# If you want to send the email now, use #deliver_now
UserMailer.welcome(@user).deliver_now

# If you want to send the email asynchronously, use #deliver_later
UserMailer.welcome(@user).deliver_later
```

The `deliver_now` and `deliver_later` methods are Action Mailer's counterparts
to `perform_now` and `perform_later`. Under the hood, `deliver_later` works by
enqueuing an `ActionMailer::MailDeliveryJob` — a built-in Active Job job that
Rails provides — which goes through the same queuing pipeline as any job you
define yourself, and will eventually call the `perform` method in your Job
class.

[`perform_now`]:
    https://api.rubyonrails.org/classes/ActiveJob/Execution.html#method-i-perform_now
[`perform_later`]:
 https://api.rubyonrails.org/classes/ActiveJob/Enqueuing/ClassMethods.html#method-i-perform_later
[`set`]:
    https://api.rubyonrails.org/classes/ActiveJob/Core/ClassMethods.html#method-i-set

> ### Supported Parameter Types for `perform`

Reviewer: It seems this guide mostly uses "argument" instead of "parameter":
```suggestion
### Supported Argument Types for `perform`
```

ActiveJob supports the following types of arguments by default:

  - Basic types (`NilClass`, `String`, `Integer`, `Float`, `BigDecimal`,
    `TrueClass`, `FalseClass`)
  - `Symbol`
  - `Date`
  - `Time`
  - `DateTime`
  - `ActiveSupport::TimeWithZone`
  - `ActiveSupport::Duration`
  - `Hash` (Keys should be of `String` or `Symbol` type)
  - `ActiveSupport::HashWithIndifferentAccess`
  - `Array`
  - `Range`
  - `Module`
  - `Class`

Active Job supports
[GlobalID](https://github.com/rails/globalid/blob/main/README.md) for

> parameters. This makes it possible to pass live Active Record objects to your

Reviewer: ```suggestion
arguments. This makes it possible to pass live Active Record objects to your
```

job instead of class/id pairs, which you then have to manually deserialize.

For example, instead of having to do something like this:

```ruby
class GuestsCleanupJob < ApplicationJob
  def perform(guests_class, guests_id, depth)
    guests = guests_class.constantize.find(guests_id)
    guest.cleanup(depth)
  end
end
```

> Using GlobedId, you can simply do:

Reviewer: ```suggestion
Using GlobalID, you can simply do:
```

end
```

```irb

> last_video = Video.last
> ProcessVideoJob.perform_later(last_video)

Reviewer: If this is an `irb` example, we should use the prefix:
```suggestion
demo(dev)> last_video = Video.last
demo(dev)> ProcessVideoJob.perform_later(last_video)
```

Author: I think I'll change it to Ruby as it doesn't show any irb output, so we don't need the extra prompt.

in line with other Active Record bulk methods. Since callbacks run on individual
jobs, they can't take advantage of the bulk nature of this method.

However, the `perform_all_later` method does fire an
[`enqueue_all.active_job`](active_support_instrumentation.html#enqueue-all-active-job)
event which you can subscribe to using `ActiveSupport::Notifications`.

The method
[`successfully_enqueued?`](https://api.rubyonrails.org/classes/ActiveJob/Core.html#method-i-successfully_enqueued-3F)
can be used to find out if a given job was successfully enqueued.

#### Queue Backend Support

For `perform_all_later`, bulk enqueuing needs to be backed by the queue backend.
Solid Queue, the default queue backend, supports bulk enqueuing using
`enqueue_all`.

[Other backends](#alternate-queuing-backends) like Sidekiq have a `push_bulk`
method, which the Sidekiq adapter users. internally to push a large number of
jobs to Redis and prevent the round trip network latency. GoodJob also supports
bulk enqueuing with the `GoodJob::Bulk.enqueue` method.

If the queue backend does *not* support bulk enqueuing, `perform_all_later` will
enqueue jobs one by one.

> ### Job Attributes

Reviewer: Job attributes are mostly (only?) [relevant for continuable jobs.](https://github.com/rails/rails/pull/57208). The example also uses the `ActiveJob::Continuable` API which hasn't been mentioned here yet and is explained later in the guide.
So I think this section would fit better in the "Job Continuations" section.

Author: Good point. I moved this section to the Job Continuables. Also added a couple lines explaining what the big example is doing.

Reviewer: That section is much clearer now!

being enqueued based on a condition:

```ruby
class GuestsCleanupJob < ApplicationJob
  before_enqueue do |job|
    throw :abort if ENV.fetch("DISABLE_GUESTS_CLEANUP_JOB", true)
  end

  def perform(guest)
    # ...
  end
end
```

When `:abort` is thrown in a `before_enqueue` callback, the job will not be
enqueued and `perform_later` will return `false`. When thrown in a
`before_perform` callback, the job will not be performed. It will also skip the
execution of any subsequent before, around and after callbacks.

NOTE: Throwing an `:abort` does not trigger `after_discard`. The `after_discard`
callback is specifically tied to the `discard_on` mechanism.

Job Continuations
-----------------

Active Job Continuations allow jobs to be split into resumable steps, so that
long-running jobs can make progress after interruptions. When using
continuations, the job automatically resumes from the last completed step,
instead of restarting from the beginning.

To use continuations, include the `ActiveJob::Continuable` module in your Job
class. You can then define each step inside the `perform` method using
[`step`](https://api.rubyonrails.org/classes/ActiveJob/Continuation/Step.html)
inside.

> 

Reviewer: The first example shows a lot of new concepts.
Maybe show a simple example here (without the cursor)?

```ruby
class ProcessImportJob < ApplicationJob
  include ActiveJob::Continuable

  def perform(import_id)
    # Always runs on job start, even when resuming from an interrupted step.
    @import = Import.find(import_id)

    # Step defined using a block
    step :initialize do
      @import.initialize
    end

    step :process do
      @import.records.find_each { |record| record.process }
    end

    # Step defined by referencing a method
    step :finalize
  end

  private
    def finalize
      @import.finalize
    end
end
```

Steps can use an optional
[cursor](https://api.rubyonrails.org/classes/ActiveJob/Continuation.html#class-ActiveJob::Continuation-label-Cursors)
to track progress *within* the step. The code in the step is responsible for
using the cursor to continue from the appropriate location after an
interruption.

Steps are executed as soon as they are encountered. Code that is not part of a
step will be executed on each job run. If a job is interrupted, previously
completed steps will be skipped. If a step is in progress, it will be restarted

> or resumed with the last recorded cursor if using cursors.  For example:

Reviewer: Before introducing cursors maybe explain how regular steps are executed first?
For example moving "Steps are executed .. steps will be skipped" sentences before "Steps can use an optional cursor...":
```suggestion
Steps are executed as soon as they are encountered. Code that is not part of a
step will be executed on each job run. If a job is interrupted, previously
completed steps will be skipped.

Steps can use an optional
[cursor](https://api.rubyonrails.org/classes/ActiveJob/Continuation.html#class-ActiveJob::Continuation-label-Cursors)
to track progress *within* the step. The code in the step is responsible for
using the cursor to continue from the appropriate location after an
interruption.

If a step is in progress, it will be restarted
or resumed with the last recorded cursor if using cursors.  For example:
```

Author: Good call. Started with the simpler example without cursor. Reworded the explanation paragraphs. And added a follow-on shorter example with cursor.

> ```ruby

Reviewer: If we show a simpler example first, we could remove some of the code from this more complex example, or only show the step with cursor.

    step :process do |step|
      @import.records.find_each(start: step.cursor) do |record|
        record.process
        step.advance!
      end
    end

    # Step defined by referencing a method
    step :finalize
  end

  private
    def finalize
      @import.finalize
    end
end
```

Each step can be declared with a block or by referencing a method name. The
block will be called with the step object as an argument. Methods can either
take no arguments or a single argument for the step object.

Job Continuations make it easier to build long-running or multi-phase jobs that
can safely pause and resume without losing progress. For more details, see
[ActiveJob::Continuation](https://api.rubyonrails.org/classes/ActiveJob/Continuation.html).

Default Backend: Solid Queue
------------------------------

Solid Queue is a database-backed queue backend for Active Job and the default
queue backend for Rails version 8.0 onwards. Rather than requiring a separate
infrastructure dependency like Redis, Solid Queue uses your existing database to
persist and process jobs. It supports delayed jobs, job priorities, concurrency
controls, recurring jobs, and bulk enqueuing.

> ### Set Up and Default Configuration

Reviewer: In other guides we use "Setup":
```suggestion
### Setup and Default Configuration
```

config.solid_queue.connects_to = { database: { writing: :queue } }
```

Additionally, the database connection for the `queue` database is configured in
`config/database.yml`:

```yaml#8
# config/database.yml
# Store production database in the storage/ directory, which by default
# is mounted as a persistent Docker volume in config/deploy.yml.
production:
  primary:
    <<: *default
    database: storage/production.sqlite3
  queue:
    <<: *default
    database: storage/production_queue.sqlite3
    migrations_paths: db/queue_migrate
```

NOTE: The key `queue` from the database configuration needs to match the key
used in the configuration for `config.solid_queue.connects_to` (as highlighted
in the code snippets above).

In order to start using Solid Queue, run `db:prepare` so your database has Solid
Queue related tables:

```bash
$ bin/rails db:prepare
```

TIP: You can find the schema for the `queue` database in `db/queue_schema.rb`,
which is generated automatically. It will contain tables like
`solid_queue_jobs`, `solid_queue_recurring_executions`,
`solid_queue_scheduled_executions`, and more.

Finally, to start the queue and start processing jobs you can run:

```bash

> bin/jobs start

Reviewer: ```suggestion
$ bin/jobs start
```

```

#### Development Environment

Rails provides an asynchronous in-process queuing backend, which keeps the jobs
in memory. With the default `async` adapter, if the process crashes or the
machine is reset, then all outstanding jobs are lost. This can be acceptable for
non-critical jobs in development.

Alternatively, you can use Solid Queue in development. It can be configured it

> in the same way as in the production environment:

Reviewer: ```suggestion
Alternatively, you can use Solid Queue in development. It can be configured
in the same way as in the production environment:
```

```ruby#3
# config/environments/development.rb
config.active_job.queue_adapter = :solid_queue
config.solid_queue.connects_to = { database: { writing: :queue } }
```

Add `queue` to the development database configuration:

> ```yaml#6

Reviewer: We can probably highlight keys under `queue` as well:
````suggestion
```yaml#6-9
````

# config/database.yml
development:
  primary:
    <<: *default
    database: storage/development.sqlite3
  queue:
    <<: *default
    database: storage/development_queue.sqlite3
    migrations_paths: db/queue_migrate
```

### Workers, Dispatchers, Supervisors

Solid Queue uses three types of processes to handle job queueing and execution.
*Workers* poll queues for jobs that are ready to run and execute them.
*Dispatchers* handle scheduled jobs — they check for jobs due to run in the
future and move them into the ready queue for workers to pick up. Both workers

> and dispatchers are managed by a *Supervisor*, which forks and monitors them.

Reviewer: Would it be clearer if we describe the types in a list?
```suggestion
Solid Queue uses three types of processes to handle job queueing and execution.
1. *Workers* poll queues for jobs that are ready to run and execute them.
2. *Dispatchers* handle scheduled jobs — they check for jobs due to run in the
future and move them into the ready queue for workers to pick up.
3. *Supervisor* manages both workers and dispatchers — it forks and monitors them.
```

Author: Yes, Done (manually)!

> When you run `bin/rails jobs`, you're starting the supervisor process, which in

Reviewer: I think Solid Queue is started by running `bin/jobs start`:
```suggestion
When you run `bin/jobs start`, you're starting the supervisor process, which in
```
https://github.com/rails/solid_queue/blob/main/lib/generators/solid_queue/install/templates/bin/jobs

Author: Thanks for catching this!

#### Queue Order

Queue order is the primary way to prioritize work in Solid Queue. The order in
which queues are listed in `config/queue.yml` for a worker determines the
polling order. A worker will not pull jobs from a lower priority queue (listed
later in the array) until all higher priority queues are empty:

```yaml
production:
  workers:
    - queues: [critical, default, low]
      threads: 5
```

With the above configuration, no jobs will be taken from the `default` queue
while the `critical` queue has jobs waiting, and no jobs will be taken from
`low` while either `critical` or `default` has jobs waiting. Solid Queue has
strict ordering (unlike other queuing backend which may allow relative weights
so that lower priority queues still receive a proportional share of processing
time). It is possible for lower queues to be starved if higher queues are
consistently busy.

NOTE: One way to name queues is based on latency. So instead of "critical",
"default", or "low", queues could be named "within_30_seconds",
"within_5_minutes", and "within_1_hour". This can be enforced like a contract by
configuring your queuing backend to notify your engineering team if a job sits

> in a given queue longer than the corresponding time.

Reviewer: This NOTE is probably interesting with other backend as well.
What do you think about moving it to the "Naming Queues" section instead?

Also, naming queues isn't required for understanding queue order. So should it be a TIP instead of a NOTE?

Author: moved and changed to TIP!

queues.

#### Retries

In Solid Queue, retries need to be configured explicitly using Active Job's
`retry_on`:

```ruby
class ExternalApiJob < ApplicationJob
  retry_on Net::TimeoutError, wait: :exponentially_longer, attempts: 5
  retry_on ActiveRecord::Deadlocked, wait: 2.seconds, attempts: 3

  def perform
    # ...
  end
end
```

This can also be set globally in `ApplicationJob` if you want a default retry
policy across all jobs. Failed jobs that aren't configured with `retry_on` will
go straight to failed executions without retrying.

### Concurrency Controls

Solid Queue extends Active Job with concurrency controls, allowing you to limit
how many jobs of a certain type can run at the same time. This is useful for
protecting shared resources, such as ensuring only one export job runs per
account at a time, or capping the number of concurrent API calls to an external
service.

Concurrency controls are declared using `limits_concurrency` in your job class:

```ruby
class InvoiceExportJob < ApplicationJob
  limits_concurrency to: 1, key: ->(account_id) { "invoice_export_#{account_id}" }, duration: 10.minutes

  def perform(account_id)
    # ...
  end
end
```

> In the above example:

Reviewer: I think this needs a linebreak:

<img width="842" height="113" alt="Image" src="https://github.com/user-attachments/assets/806c7092-7afa-43fa-806a-403fde874b45" />

```suggestion
In the above example:

```

- The `to` option sets the maximum number of jobs that can run concurrently.
- The `key` lambda computes a concurrency key from the job's arguments. In the
  example above, the limit of 1 applies per account rather than globally.
- The `duration` option acts as a failsafe. So if a worker dies mid-job and
  fails to release its lock, any blocked jobs become candidates for release once

>   duration has elapsed.

Reviewer: In the next paragraph options are prepended with a colon (mostly):
```suggestion
- The `:to` option sets the maximum number of jobs that can run concurrently.
- The `:key` lambda computes a concurrency key from the job's arguments. In the
  example above, the limit of 1 applies per account rather than globally.
- The `:duration` option acts as a failsafe. So if a worker dies mid-job and
  fails to release its lock, any blocked jobs become candidates for release once
  duration has elapsed.
```

When a job with concurrency controls is enqueued, Solid Queue checks a
database-backed lock for the computed key. If the lock is available, the job is

> marked ready for execution. If not, the behavior depends on the `on_conflict:`

Reviewer: ```suggestion
marked ready for execution. If not, the behavior depends on the `:on_conflict`
```

option. If `on_conflict` is set to `:block` (the default), the job is held in a
blocked state and marked ready only when a running job finishes. The other
option is `:discard`, in which case the job is dropped entirely.

> You can also scope limits across *different* job classes using the `group:`

Reviewer: ```suggestion
You can also scope limits across *different* job classes using the `:group`
```

option:

```ruby
class AnalyticsExportJob < ApplicationJob
  limits_concurrency to: 1, key: ->(account_id) { account_id }, group: "account_exports", duration: 10.minutes
end

class InvoiceExportJob < ApplicationJob
  limits_concurrency to: 1, key: ->(account_id) { account_id }, group: "account_exports", duration: 10.minutes
end
```

In the above example, both job classes share the same concurrency limit per the
"account_exports" group, which means only one export of either type will run at
a time for a given account.

NOTE: Concurrency controls do carry overhead since blocked executions must be
tracked and locks created and updated. So they should be used sparingly. For
simple throughput limiting, constraining the number of worker threads per queue
is more efficient.

> WARN: Concurrency controls are not compatible with bulk enqueuing via

Reviewer: This is currently not getting picked up as a warning:

<img width="853" height="74" alt="Image" src="https://github.com/user-attachments/assets/1b016bb0-45a5-496b-80e4-9fbc672ab2b0" />

```suggestion
WARNING: Concurrency controls are not compatible with bulk enqueuing via
```

Redis-backed backends where this problem doesn't arise in the same form. If
you're unsure whether your code relies on transactional integrity, enabling
`enqueue_after_transaction_commit` globally in `ApplicationJob` is the safest
default.

You can read more about [Transactional Integrity in the Solid Queue
documentation](https://github.com/rails/solid_queue?tab=readme-ov-file#jobs-and-transactional-integrity)

### Recurring Tasks

Solid Queue supports recurring tasks, similar to cron jobs. These tasks are
defined in a configuration file (by default, `config/recurring.yml`) and can be
scheduled at specific times. Here's an example of a task configuration:

```yaml
production:
  a_periodic_job:
    class: MyJob
    args: [42, { status: "custom_status" }]
    schedule: every second
  a_cleanup_task:
    command: "DeletedStuff.clear_all"
    schedule: every day at 9am
```

Each task specifies a `class` or `command` and a `schedule` (parsed using
[Fugit](https://github.com/floraison/fugit)). You can also pass arguments to
jobs, such as in the example for `MyJob` where `args` are passed. This can be
passed as a single argument, a hash, or an array of arguments that can also

> include key word arguments as the last element in the array.

Reviewer: ```suggestion
include keyword arguments as the last element in the array.
```

You can learn more about [Recurring
Tasks](https://github.com/rails/solid_queue?tab=readme-ov-file#recurring-tasks)
in the Solid Queue documentation.

Alternate Queuing Backends
--------------------------

While Solid Queue is the default queuing backend in Rails, Active Job is
designed to work seamlessly with different queuing backends. Switching to an
alternative backend, such as [Sidekiq](https://github.com/sidekiq/sidekiq),
[GoodJob](https://github.com/bensheldon/good_job), or
[Resque](https://github.com/resque/resque), requires only a configuration change
(typically with no modifications to your job code), along with adding the
queuing backend's adapter to your Gemfile.

You can get an up-to-date list of the adapters in the the API Documentation for
[`ActiveJob::QueueAdapters`](https://api.rubyonrails.org/classes/ActiveJob/QueueAdapters.html).

> 

Reviewer: All the build-in adapters for external gems have been [deprecated](https://github.com/search?q=repo%3Arails%2Frails+activejob+deprecated&type=pullrequests), so we should probably remove this as the list won't list external adapters in the future:
```suggestion
```

Another option would be showing the list of alternative backends here, instead of at the end of this section.

Author: Good call! Moving the list of backends up here.

process to be started alongside your Rails application, and some (like Sidekiq)
require additional infrastructure such as Redis.

NOTE: Switching backends doesn't migrate jobs already sitting in the old queue.
You'll need to drain the old queue before switching, or run both backends in
parallel temporarily to let existing jobs complete.

TIP: If you use `config.active_job.queue_name_prefix`, make sure your new
backend's worker configuration listens to the prefixed queue names, not the bare
names.

Here is a noncomprehensive list of documentation:

- [Sidekiq](https://github.com/mperham/sidekiq/wiki/Active-Job)
- [Resque](https://github.com/resque/resque/wiki/ActiveJob)
- [Sneakers](https://github.com/jondot/sneakers/wiki/How-To:-Rails-Background-Jobs-with-ActiveJob)
- [Queue Classic](https://github.com/QueueClassic/queue_classic#active-job)
- [Delayed Job](https://github.com/collectiveidea/delayed_job#active-job)
- [Que](https://github.com/que-rb/que#additional-rails-specific-setup)
- [Good Job](https://github.com/bensheldon/good_job#readme)

Monitoring and Handling Failed Jobs
-----------------------------------

### Monitoring With Mission Control

> The [Mission Control](https://github.com/rails/mission_control-jobs) tool is a

Reviewer: Maybe "gem" or "engine" is more descriptive?
```suggestion
The [Mission Control](https://github.com/rails/mission_control-jobs) engine is a
```

end
```

These class-level methods also accept a block, which works well when the
callback logic is short enough to fit on a single line. For example, sending
metrics for every enqueued job:

```ruby
class ApplicationJob < ActiveJob::Base
  before_enqueue { |job| Rails.logger.info "Enqueuing #{job.class.name}" }
end
```

### Available Callbacks

> There are several other callbacks that Active Job supports.

Reviewer: We just mentioned `before_enqueue` above, so maybe "other" is a bit confusing here when we start with `before_enqueue`:
```suggestion
There are several callbacks that Active Job supports.
```

[`retry_on`]:
    https://api.rubyonrails.org/classes/ActiveJob/Exceptions/ClassMethods.html#method-i-retry_on

> ### Missing Records

Reviewer: The previous section already mentions `ActiveJob::DeserializationError` in the `discard_on` example. Not sure if we should maybe use a different error there, or maybe we can merge these two sections?

Author: Changed the first one to Net::OpenTimeout, as that's a better example anyway.

> GlobalID allows serializing full Active Record objects passed to `#perform`.

Reviewer: This seems to imply the AR objects are serialized with all their attributes.
Maybe something like:
```suggestion
GlobalID will use the unique identifier to locate the full Active Record object when calling `#perform`.
```
